tslearn/nonmyopic.py

```diff
@@ -0,0 +1,312 @@
+                self.pyck_[current_classe, k] = (
+                    self.pyck_[current_classe, k] / current_sum
+                )
+
+        for t in range(self.minimum_time_stamp, X.shape[1] + 1):
+            self.classifiers_[t].fit(X1[:, :t], y1)
+            for k in range(0, self.n_clusters):
+                index_cluster = np.where(c_k2 == k)
+
+                if len(index_cluster[0]) != 0:
+                    X2_current_cluster = np.squeeze(
+                        X2[index_cluster, :t], axis=0
+                    )
+                    y2_current_cluster = y2[tuple(index_cluster)]
+                    y2_hat = self.classifiers_[t].predict(
+                        X2_current_cluster[:, :t]
+                    )
+                    conf_matrix = confusion_matrix(
+                        y2_current_cluster, y2_hat, labels=y_classes_indices, normalize="true"
+                    )
+                    np.fill_diagonal(conf_matrix, 0)
+                    self.pyhatyck_["pyhatycks{0}".format(k)].append(
+                        conf_matrix
+                    )
+
+    def get_cluster_probas(self, X):
+        """
+        This function computes the probabilities of the time series xt to be in a cluster of the model
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated
+
+        Returns
+        -------
+        vector : the probabilities of the given time series to be in the clusters of the model
+        """
+        diffs = X[np.newaxis, :] - self.cluster_.cluster_centers_[:, len(X)]
+        distances_clusters = np.linalg.norm(diffs, axis=2)
+        distances_clusters = np.asarray(distances_clusters)
+        minimum_distance = np.min(distances_clusters)
+        average_distance = np.mean(distances_clusters)
+        minimum_distance_to_average = (
+            average_distance - minimum_distance
+        ) / average_distance
+        delta = (average_distance - distances_clusters) / average_distance
+        probas = 1.0 / (
+            np.exp(self.lamb * minimum_distance_to_average)
+            + np.exp(self.lamb * (minimum_distance_to_average - delta))
+            + 1e-6
+        )
+        return probas / probas.sum()
+
+    def _expected_cost(self, X, tau):
+        """
+        From a incomplete series xt, this function compute the expected cost of a prediction made at time "last time of
+        xt + tau"
+
+        Parameters
+        ----------
+        X: vector
+            a time series that may be truncated for which we want to have the cost
+        tau: int
+            gives the future time for which the cost is calculated
+
+        Returns
+        --------
+        float : the computed cost
+        """
+        cost = 0
+        proba_clusters = self.get_cluster_probas(X=X)
+        truncated_t = X.shape[-1]
+        for k in range(0, self.n_clusters):
+            sum_pyhatyck = np.sum(self.pyhatyck_["pyhatycks{0}".format(k)][
```

Comment:
My comment above means we should have `self.pyhatyck_` that would be an array of shape (n_classes, n_classes, n_clusters) and the operation you are performing in this for loop could then be vectorized for efficiency purpose.